Add first solution4-1

Nigma-Ks opened the pull request.

Nigma-Ks committed:
Add first solution4-1

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+    {
+        static async Task Main(string[] args)
+        {
+            string path = "..\\..\\..\\..\\hw4_1.txt";
```
По хорошему, программа должна принимать путь к файлу в качестве параметра.

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+            string expression;
+            using (StreamReader reader = new StreamReader(path))
+            {
+                if ((expression = await reader.ReadToEndAsync()) == null)
```
```suggestion
                if ((expression = reader.ReadToEnd()) == null)

```

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+                if ((expression = await reader.ReadToEndAsync()) == null)
+                {
+                    Console.WriteLine("Empty file!");
+                    reader.Close();
```
```suggestion
```
При выходе из секции `using` вызовется метод `Dispose`. Причем не важно, по какой причине мы вышли из этой секции. Даже в случае исключения или `return`. Так что закрывать файл вручную - излишне.

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+                return;
+            }
+
+            ExpressionTree newCorrectTree = new ExpressionTree(expression);
```
Зачем строить дерево дважды?

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+            }
+            
+            Console.WriteLine("Expression in file: " + expression);
+            try
+            {
+                ExpressionTree tree = new ExpressionTree(expression);
+            }
```
```suggestion
            ExpressionTree tree;

            try
            {
                 tree = new ExpressionTree(expression);
            }
```
В этом случае дерево будет доступно после секции `try-catch`.

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+        }
+    }
+
+    public class ExpressionTree
```
Классы должны быть в отдельных файлах.

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+            string rightBracketExpression = ConstructRightBracketsExpression();
+            expression = expression.Replace(" ", "");
+            return String.Compare(expression, rightBracketExpression) == 0;
+        }
```
```suggestion
        }

```

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+        public ExpressionTree(string expression)
+        {
+            int currOperand;
+            bool notEmptyStack;
```
Неиспользуемая переменная

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+                    }
+                    catch (InvalidOperationException e)
+                    {
+                        Console.WriteLine("File was incorrect!");
+                        Console.WriteLine("Message: " + e.Message);
+                        Console.WriteLine("StackTrace:\n" + e.StackTrace);
```
Печатать должен только `Main`

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+                    }
+                    catch (InvalidOperationException e)
+                    {
+                        Console.WriteLine("Message: " + e.Message);
+                        Console.WriteLine("StackTrace:\n" + e.StackTrace);
```
Только в `Main`

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+                {
+                    WrongSymbolsInExpressionException e =
+                        new WrongSymbolsInExpressionException("Incorrect numbers in file!");
+                    Console.WriteLine("Message: " + e.Message);
+                    Console.WriteLine("StackTrace:\n" + e.StackTrace);
```
В `Main`

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+            }
+            catch (InvalidOperationException e)
+            {
+                Console.WriteLine("File was incorrect!");
+                Console.WriteLine("Message: " + e.Message);
+                Console.WriteLine("StackTrace:\n" + e.StackTrace);
```
В `Main`

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+        {
+            public string Operation;
+            public OperandVertex LeftOperand;
+            public OperandVertex RightOperand;
```
```suggestion
            public IVertex LeftOperand;
            public IVertex RightOperand;
```
Неудачная иерархия классов.
Предполагалось, что получится интерфейс или абстрактный класс, от которого наследуются операции и операнды. Тогда в операции можно хранить операнды единообразно, вне зависимости от того, это числа или другие операции.

Lozov-Petr commented on hw4_1/hw4_1/Program.cs:
```diff
@@ -0,0 +1,300 @@
+        public class OperandVertex
+        {
+            public int Value;
+            public OperationVertex? Operation;
```
Вот такого быть не должно. Операнд должен знать только о числе.

Lozov-Petr reviewed (commented):
2.5

Nigma-Ks committed:
1. Теперь путь к файлу принимается в качестве параметра
2. Изменила способ считывания текста
3. Дерево больше не строится дважды
4. Сообщение об ошибке в main
5. Исправила классы вершин, добавила интерфейс

Nigma-Ks committed:
Переместила описания классов в новый файлы

Nigma-Ks committed:
Класс ошибок больше не пустой
